Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/cache/FileCacheDownloader.kt

```diff
@@ -0,0 +1,385 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package com.github.adamantcheese.chan.core.cache
+
+import android.os.Handler
+import android.os.Looper
+import androidx.annotation.AnyThread
+import androidx.annotation.MainThread
+import androidx.annotation.WorkerThread
+import com.github.adamantcheese.chan.Chan
+import com.github.adamantcheese.chan.core.di.NetModule
+import com.github.adamantcheese.chan.utils.BackgroundUtils
+import com.github.adamantcheese.chan.utils.Logger
+import com.github.k1rakishou.fsaf.FileManager
+import com.github.k1rakishou.fsaf.file.RawFile
+import okhttp3.Call
+import okhttp3.OkHttpClient
+import okhttp3.Request
+import okhttp3.ResponseBody
+import okhttp3.internal.closeQuietly
+import okio.*
+import java.io.Closeable
+import java.io.IOException
+import java.io.OutputStream
+import java.util.*
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.atomic.AtomicInteger
+
+class FileCacheDownloader(
+        private val fileManager: FileManager,
+        // Main thread only.
+        private val callback: Callback?,
+        private val output: RawFile,
+        val url: String
+) : Runnable {
+    private val handler: Handler = Handler(Looper.getMainLooper())
+    private val listeners = ArrayList<FileCacheListener>()
+
+    // Main and worker thread.
+    private val running = AtomicBoolean(false)
+    private val cancel = AtomicBoolean(false)
+
+    // When encountering an IOException during the attempt to download a file instead of immediately
+    // terminating the downloading we will try to download it couple of more times (3 times max).
+    // If we are still unable to download the file after all of the attempts (no internet connection)
+    // we give up and throw the latest exception. This should fix the annoying "Saved image failed"
+    // error when downloading thread albums.
+    //
+    // The reason is actually SocketTimeoutException is the server forcefully drops
+    // the connection if it detects that we are downloading files?
+    private val retryAttempts = AtomicInteger(3)
+    private val retryTimeoutsSeconds = arrayOf(5L, 2L, 1L)
+
+    @MainThread
+    fun addListener(callback: FileCacheListener) {
+        BackgroundUtils.ensureMainThread()
+
+        listeners.add(callback)
+    }
+
+    /**
+     * Cancel this download.
+     */
+    @MainThread
+    fun cancel() {
+        BackgroundUtils.ensureMainThread()
+
+        if (cancel.compareAndSet(false, true)) {
+            // Did not start running yet, mark finished here.
+            if (!running.get() && callback != null) {
+                Thread.currentThread().interrupt();
```

Comment:
Uhhh, this may actually not be good because this will interrupt the main thread.

Reply:
Gonna fix this today, do not merge this one yet.